src/ripple/nodestore/impl/DeterministicShard.cpp

```diff
@@ -0,0 +1,256 @@
+        return false;
+    }
+
+    Config const& config {app_.config()};
+
+    Section section {config.section(ConfigSection::shardDatabase())};
+    std::string const type {get<std::string>(section, "type", "nudb")};
+
+    if (type != "nudb")
+    {
+        JLOG(j_.error()) <<
+                "shard " << index_ <<
+                " backend type " << type << " not supported";
+        return false;
+    }
+
+    auto factory {Manager::instance().find(type)};
+    if (!factory)
+    {
+        JLOG(j_.error()) <<
+                "shard " << index_ <<
+                " failed to create factory for backend type " << type;
+        return false;
+    }
+
+    ctx_->start();
+
+    section.set("path", tempdir_.string());
+    backend_ = factory->createInstance(
+        NodeObject::keyBytes, section, scheduler_, *ctx_, j_);
+
+    if (!backend_)
+    {
+        JLOG(j_.error()) <<
+                "shard " << index_ <<
+                " failed to create backend type " << type;
+        return false;
+    }
+
+    // Open or create the NuDB key/value store
+    bool preexist = exists(tempdir_);
+    if (preexist) {
+        remove_all(tempdir_);
+        preexist = false;
+    }
+
+    backend_->open(!preexist, digest(0));
+    backend_->close();
+
+    {
+        boost::filesystem::path keypath = tempdir_ / "nudb.key";
+        boost::filesystem::path datpath = tempdir_ / "nudb.dat";
+
+        class stream_ : public std::fstream {
+            std::streampos size_;
+        public:
+            stream_(const std::string &s)
+                : std::fstream(s, std::ios_base::in | std::ios_base::out | std::ios_base::binary)
+                , size_ (0)
+            {
+            }
+            void read(size_t pos, unsigned char *buf, size_t size, boost::system::error_code &ec)
+            {
+                std::fstream::seekg(pos);
+                std::fstream::read((char *)buf, size);
+                size_ = std::fstream::tellg() - (std::streampos)pos;
+                if (size != size_)
+                {
+                    ec = boost::system::errc::make_error_code(boost::system::errc::io_error);
+                }
+            }
+            void write(size_t pos, unsigned char *buf, size_t size, boost::system::error_code &ec)
+            {
+                std::fstream::seekp(pos);
+                std::fstream::write((char *)buf, size);
+                size_ = std::fstream::tellp() - (std::streampos)pos;
+                if (size != size_)
+                {
+                    ec = boost::system::errc::make_error_code(boost::system::errc::io_error);
+                }
+            }
+            size_t size(boost::system::error_code &ec)
+            {
+                return size_;
+            }
+        };
+
+        stream_ keyfile(keypath.string());
+        stream_ datfile(datpath.string());
+
+        if (!keyfile || !datfile)
+            return false;
+
+        nudb::detail::key_file_header keyhead;
```

Comment:
@p2peer, @miguelportilla we are reaching inside `NuDB` internals here; we absolutely **do not** want to do this. We need to modify `NuDB` so that it cleanly supports the operations that we need to perform.

Reply:
Which way do you want to modify NuDB: 1) include NuDB to rippled and change the code; 2) connect to maintainers of NuDB and ask them to push a modification? I don't know if the way 2 will be effective.

Reply:
Author of NuDB said that he will research this, see his comment below. While he is researching, I put code under macro NUDB_SUPPORTS_CUSTOM_UID.